### PR TITLE
Selector accept int string

### LINKED ISSUE #2891: `selector` should accept integers and strings, for all functions accepting the `selector` argument
Example functions are `update_trace`, `for_each_shape`, etc.

If the `selector` keyword argument is an integer, it should behave as follows:
 - If row and col aren't specified, it should act as an index into `fig.data`, i.e., `fig.data[selector]`
 - If they are, then it acts as an index into the objects on that subplot, e.g., `list(fig.select_traces(row=r,col=c))[selector])`

If the `selector` keyword argument is a string, it should create a dictionary whose sole key is `type` and `selector` is the value at the key, e.g., `if type(selector) == type(str()): selector = dict(type=selector)`.

### PR DESCRIPTION
**Pull request by nicholas-esterer**
closes #2891

### CODE REVIEW

**Comment by nicholas-esterer** [comment]
Do we want something like:

`fig.select_annotations(selector='idunno')`

to raise ValueError or just return an empty generator?

**Comment by nicolaskruchten** [comment]
I think a ValueError or whatever the equivalent error you'd get from `selector=dict(type="idunno")`

**Comment by nicholas-esterer** [comment]
That doesn't throw an error either, it also just returns an empty generator. So throwing an error in this case might actually break the API.

**Comment by nicholas-esterer** [comment]
So other than the uncertainty surrounding raising an error or not, this PR can be reviewed :+1:

**Comment by nicolaskruchten** [comment]
@jonmmease can you give this PR a sanity-check, implementation-wise, please?

**Comment by nicolaskruchten** [comment]
The behaviour seems like what I want :)

**Review by jonmmease — COMMENTED** [review]
Functionality and tests look good!

I left a few comments/suggestions on design.

File: packages/python/plotly/plotly/basedatatypes.py [hunk 1]
```diff
@@ -979,6 +988,34 @@ def _selector_matches(obj, selector):
...
+        If selector is an integer, the resulting sequence obtained after
+        sucessively filtering by each function in funcs is indexed by this
+        integer.
+        Otherwise selector is used as the selector argument to
+        self._selector_matches which is used to filter down the sequence.
+        The function returns the sequence (an iterator).
+        """
+
+        # if selector is not an int, we call it on each trace to test it for selection
+        if type(selector) != type(int()):
```
[the remark this answers is not in the document]
Comment (nicolaskruchten) [reply packages/python/plotly/plotly/basedatatypes.py]: done, along with all the other similar issues flagged by flake8

File: packages/python/plotly/plotly/basedatatypes.py [hunk 2]
```diff
@@ -934,23 +941,25 @@ def select_traces(self, selector=None, row=None, col=None, secondary_y=None):
...
-            yield trace
+        return _generator(self._filter_by_selector(self.data, funcs, selector))
 
     @staticmethod
     def _selector_matches(obj, selector):
         if selector is None:
             return True
+        # If selector is a string then put it at the 'type' key of a dictionary
+        # to select objects where "type":selector
+        if type(selector) == type(str()):
```
[the remark this answers is not in the document]
Comment (nicolaskruchten) [reply packages/python/plotly/plotly/basedatatypes.py]: done

**Pull request merged**